File: factory-method/factory.kt
abstract class Animal {
    abstract fun speak(): String
}

class Dog : Animal() {
    override fun speak() = "Woof!"
}

class Cat : Animal() {
    override fun speak() = "Meow!"
}

class AnimalFactory {
    fun createAnimal(animalType: String): Animal? {
        return when (animalType) {
            "dog" -> Dog()
            "cat" -> Cat()
            else -> null
        }
    }
}

fun main() {
    val factory = AnimalFactory()
    val dog = factory.createAnimal("dog")
    val cat = factory.createAnimal("cat")

    dog?.let { println(it.speak()) }
    cat?.let { println(it.speak()) }
}

// kotlinc factory.kt -include-runtime -d factory.jar
// java -jar factory.jar

File: factory-method/factory.cs
using System;

abstract class Animal {
    public abstract string Speak();
}

class Dog : Animal {
    public override string Speak() {
        return "Woof!";
    }
}

class Cat : Animal {
    public override string Speak() {
        return "Meow!";
    }
}

class AnimalFactory {
    public Animal CreateAnimal(string animalType) {
        if (animalType == "dog") return new Dog();
        else if (animalType == "cat") return new Cat();
        throw new ArgumentException("Unknown animal type");
    }
}

class Program {
    static void Main(string[] args) {
        AnimalFactory factory = new AnimalFactory();
        Animal dog = factory.CreateAnimal("dog");
        Animal cat = factory.CreateAnimal("cat");

        if (dog != null) Console.WriteLine(dog.Speak());
        if (cat != null) Console.WriteLine(cat.Speak());
    }
}

// 실행 코드
// dotnet new console -n factoryApp --force
// mv factory.cs factoryApp/Program.cs
// cd factoryApp
// dotnet run